### DISCUSSION TITLE
Permutation

### DISCUSSION THREAD

**Discussion by fuzhibin** [type-challenges/type-challenges#8639] [Q&A]
type Permutation<T,U = T> = T extends never ? []:T extends T ? [T,...Permutation<Exclude<U,T>>]:never;
和
type Permutation<T,U = T> = [T] extends [never] ? []:T extends T ? [T,...Permutation<Exclude<U,T>>]:never;
的区别是什么？？？

**Comment by humandetail** [comment]
let E1: string & never extends never ? true : false // true
let E2: [string & never] extends never ? true : false // false
